File: src/js/src/utils/PathValidator.ts
import * as path from 'path';
import * as fs from 'fs';
import { InvalidFilePathError } from '../errors/FileReadingErrors';

/**
 * Utility functions for validating file paths
 */
export class PathValidator {
  /**
   * Validates that the provided file path is valid
   * @param filePath The file path to validate
   * @returns True if the path is valid, false otherwise
   */
  static isValidPath(filePath: string | null | undefined): boolean {
    if (!filePath || filePath.trim().length === 0) {
      return false;
    }

    // Check for null bytes
    if (filePath.includes('\0')) {
      return false;
    }

    // Check for empty path segments
    const segments = filePath.split(path.sep);
    if (segments.some(segment => segment.trim().length === 0 && segment !== '')) {
      return false;
    }

    // Check if path is too long (Windows MAX_PATH limit is 260, but we'll be more lenient)
    if (filePath.length > 4096) {
      return false;
    }

    // Try to normalize the path - if it throws, the path is invalid
    try {
      path.normalize(filePath);
    } catch {
      return false;
    }

    return true;
  }

  /**
   * Validates the file path and throws an error if invalid
   * @param filePath The file path to validate
   * @throws {InvalidFilePathError} When the file path is invalid
   */
  static validatePathOrThrow(filePath: string | null | undefined): void {
    if (!this.isValidPath(filePath)) {
      throw new InvalidFilePathError(filePath || '');
    }
  }

  /**
   * Checks if a file exists at the specified path
   * @param filePath The file path to check
   * @returns True if the file exists, false otherwise
   */
  static fileExists(filePath: string): boolean {
    if (!this.isValidPath(filePath)) {
      return false;
    }

    try {
      return fs.existsSync(filePath) && fs.statSync(filePath).isFile();
    } catch {
      return false;
    }
  }
}

File: src/dotnet/ApiSdk/PathValidator.cs
using System.Text.RegularExpressions;
using ApiSdk.Exceptions;

namespace ApiSdk;

/// <summary>
/// Utility class for validating file paths
/// </summary>
public static class PathValidator
{
    private static readonly char[] InvalidPathChars = Path.GetInvalidPathChars();
    private static readonly char[] InvalidFileNameChars = Path.GetInvalidFileNameChars();

    /// <summary>
    /// Validates that the provided file path is valid
    /// </summary>
    /// <param name="filePath">The file path to validate</param>
    /// <returns>True if the path is valid, false otherwise</returns>
    public static bool IsValidPath(string? filePath)
    {
        if (string.IsNullOrWhiteSpace(filePath))
        {
            return false;
        }

        // Check for invalid path characters
        if (filePath.IndexOfAny(InvalidPathChars) >= 0)
        {
            return false;
        }

        // Reject whitespace-only segments (e.g. "a/ /b"), but allow a truly empty
        // leading segment so absolute paths like "/Users/.../file.json" are valid.
        var segments = filePath.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
        if (segments.Any(s => s.Length != 0 && string.IsNullOrWhiteSpace(s)))
        {
            return false;
        }

        // Check if path is too long
        if (filePath.Length > 4096)
        {
            return false;
        }

        return true;
    }

    /// <summary>
    /// Validates the file path and throws an exception if invalid
    /// </summary>
    /// <param name="filePath">The file path to validate</param>
    /// <exception cref="InvalidFilePathException">Thrown when the file path is invalid</exception>
    public static void ValidatePathOrThrow(string? filePath)
    {
        if (!IsValidPath(filePath))
        {
            throw new InvalidFilePathException(filePath ?? string.Empty);
        }
    }

    /// <summary>
    /// Checks if a file exists at the specified path
    /// </summary>
    /// <param name="filePath">The file path to check</param>
    /// <returns>True if the file exists, false otherwise</returns>
    public static bool FileExists(string filePath)
    {
        if (!IsValidPath(filePath))
        {
            return false;
        }

        return File.Exists(filePath);
    }
}

